Source/WebInspectorUI/UserInterface/Controllers/JavaScriptRuntimeCompletionProvider.js

```diff
@@ -372,30 +374,36 @@ WI.JavaScriptRuntimeCompletionProvider = class JavaScriptRuntimeCompletionProvid
                         property = quoteUsed + property.escapeCharacters(quoteUsed + "\\") + (suffix !== quoteUsed ? quoteUsed : "");
                 }
 
-                if (!property.startsWith(prefix) || knownCompletions.has(property))
+                if (knownCompletions.has(property))
+                    continue;
+
+                let startsWithPrefix = caseSensitiveMatching ? property.startsWith(prefix) : property.toLowerCase().startsWith(prefixLowerCase);
+                if (!startsWithPrefix)
                     continue;
 
                 completions.push(property);
                 knownCompletions.add(property);
             }
 
-            function compare(a, b)
+            let completionSortCriteria = new Map;
+            for (let completion of completions) {
+                completionSortCriteria.set(completion, {
+                    numericValue: isNaN(completion) ? Infinity : +completion,
+                    isRareProperty: completion.startsWith("__") && completion.endsWith("__"),
+                    startsWithPrefix: completion.startsWith(prefix),
+                });
+            }
+
+            function compare(s, t)
             {
-                // Try to sort in numerical order first.
-                let numericCompareResult = a - b;
-                if (!isNaN(numericCompareResult))
-                    return numericCompareResult;
-
-                // Sort __defineGetter__, __lookupGetter__, and friends last.
-                let aRareProperty = a.startsWith("__") && a.endsWith("__");
-                let bRareProperty = b.startsWith("__") && b.endsWith("__");
-                if (aRareProperty && !bRareProperty)
-                    return 1;
-                if (!aRareProperty && bRareProperty)
-                    return -1;
-
-                // Not numbers, sort as strings.
-                return a.extendedLocaleCompare(b);
+                let a = completionSortCriteria.get(s);
+                let b = completionSortCriteria.get(t);
+                return (
```

Comment:
Style: the `(` `)` are unnecessary
```js
return a.numericValue - b.numericValue
    || a.isRareProperty - b.isRareProperty
    || b.startsWithPrefix - a.startsWithPrefix
    || s.extendedLocaleCompare(t);
```

also should it be `a.startsWithPrefix - b.startsWithPrefix` instead?

Reply:
oh i also prefer the shorter form :)

it just might be nice to re-add some of the comments for clarity
```js
return a.numericValue - b.numericValue // Try to sort in numerical order first.
    || a.isRareProperty - b.isRareProperty // Sort __defineGetter__, __lookupGetter__, and friends last.
    || b.startsWithPrefix - a.startsWithPrefix // Prefer completions that directly match the typed text. 
    || s.extendedLocaleCompare(t); // Not numbers, sort as strings.
```

btw can we write tests for this?  perhaps we could add another test case to `LayoutTests/inspector/console/js-completions.html`?